### ISSUE TITLE
Bug: Distributed agents in JAX may be initialized with the same random key.

### ISSUE THREAD

**Issue by ethanluoyc** [deepmind/acme#136]
Hi,

I am currently implementing some distributed JAX agents in JAX following the implementation of TF D4PG agents. I am adapting the builder for this. However, I noticed that using the interface results in distributed actors being initialized with the same random key.

Here is a snippet for reproducing the problem.

```python
from launchpad.nodes.python.local_multi_processing import PythonProcess
from complr.research.skills.lp_run import main
import haiku as hk
import launchpad as lp
import jax.numpy as jnp
import jax


class Actor:
    def __init__(self, random_key) -> None:
        self._random_key = random_key

    def run(self):
        print(self._random_key)


class DistributedAgent:
    def __init__(self, seed):
        self._key = jax.random.PRNGKey(seed)

    def actor(self):
        key, self._key = jax.random.split(self._key)
        return Actor(key)

    def build(self):
        program = lp.Program(name="program")

        with program.group("actor"):
            for _ in range(4):
                _ = program.add_node(lp.CourierNode(self.actor))
        return program


def main():
    agent = Agent(0)
    program = agent.build()
    local_resources = {
        "actor": PythonProcess(env={"CUDA_VISIBLE_DEVICES": ""}),
    }
    lp.launch(
        program,
        lp.LaunchType.LOCAL_MULTI_PROCESSING,
        terminal="current_terminal",
        local_resources=local_resources,
    )


if __name__ == "__main__":
    main()
```
From the terminal, you will see the following result (filtered irrelevant log messages).

> [actor/0] I0709 00:07:24.056859 140666910336832 courier_utils.py:51] Binding: run
> [actor/0] [4146024105  967050713]
> [actor/2] I0709 00:07:24.108036 140161567921984 courier_utils.py:51] Binding: run
> [actor/1] I0709 00:07:24.108819 139808703252288 courier_utils.py:51] Binding: run
> [actor/2] [4146024105  967050713]
> [actor/1] [4146024105  967050713]
> [actor/3] I0709 00:07:24.169437 140157948270400 courier_utils.py:51] Binding: run
> [actor/3] [4146024105  967050713]

From the output you can see that all actors are initialized with the same random key, despite us splitting the key in the `actor` method. I think the issue is because LaunchPad will (cloud)pickle the constructor and recreate that in the subprocesses. However, the pickling is lazy and happens _before_ the method is called and consequently the pickle captures the seed in the builder before the splitting begins. 

It would perhaps be useful to revisit the commit f788e7574e1e6102e81069cf5fe1ef299a95a08d which introduced these changes.

**Comment by sinopalnikov** [comment]
Hi Yicheng,

Your agent could be updated as follows to pass different keys to different actor nodes:

```
from launchpad.nodes.python.local_multi_processing import PythonProcess
from complr.research.skills.lp_run import main
import haiku as hk
import launchpad as lp
import jax.numpy as jnp
import jax


class Actor:
    def __init__(self, random_key) -> None:
        self._random_key = random_key

    def run(self):
        print(self._random_key)


class DistributedAgent:
    def __init__(self, seed):
        self._seed = seed

    def actor(self, key):
        return Actor(key)

    def build(self):
        program = lp.Program(name="program")

        key = jax.random.PRNGKey(seed)

        with program.group("actor"):
            for _ in range(4):
                actor_key, key = jax.random.split(key)
                _ = program.add_node(lp.CourierNode(self.actor, actor_key))
        return program


def main():
    agent = Agent(0)
    program = agent.build()
    local_resources = {
        "actor": PythonProcess(env={"CUDA_VISIBLE_DEVICES": ""}),
    }
    lp.launch(
        program,
        lp.LaunchType.LOCAL_MULTI_PROCESSING,
        terminal="current_terminal",
        local_resources=local_resources,
    )


if __name__ == "__main__":
    main()
```

Please let me know if it doesn't work for you.

Thanks,
Danila

**Comment by ethanluoyc** [comment]
Hi Danila,

Yeah that's exactly what did in the end. I realize that for me to write the distributed agents, they may need to be structured slightly differently compared to in tf.
